cpp/src/arrow/util/compression_bz2.cc

```diff
@@ -0,0 +1,259 @@
+
+#include <bzlib.h>
+
+#include "arrow/status.h"
+#include "arrow/util/logging.h"
+#include "arrow/util/macros.h"
+
+namespace arrow {
+namespace util {
+
+constexpr int kBZ2DefaultCompressionLevel = 9;
+
+// Max number of bytes the bz2 APIs accept at a time
+static constexpr auto kSizeLimit =
+    static_cast<int64_t>(std::numeric_limits<unsigned int>::max());
+
+Status BZ2Error(const char* prefix_msg, int bz_result) {
+  DCHECK(bz_result != BZ_OK && bz_result != BZ_RUN_OK && bz_result != BZ_FLUSH_OK &&
+         bz_result != BZ_FINISH_OK && bz_result != BZ_STREAM_END);
+  StatusCode code;
+  std::stringstream ss;
+  ss << prefix_msg;
+  switch (bz_result) {
+    case BZ_CONFIG_ERROR:
+      code = StatusCode::UnknownError;
+      ss << "bz2 library improperly configured (internal error)";
+      break;
+    case BZ_SEQUENCE_ERROR:
+      code = StatusCode::UnknownError;
+      ss << "wrong sequence of calls to bz2 library (internal error)";
+      break;
+    case BZ_PARAM_ERROR:
+      code = StatusCode::UnknownError;
+      ss << "wrong parameter to bz2 library (internal error)";
+      break;
+    case BZ_MEM_ERROR:
+      code = StatusCode::OutOfMemory;
+      ss << "could not allocate memory for bz2 library";
+      break;
+    case BZ_DATA_ERROR:
+      code = StatusCode::IOError;
+      ss << "invalid bz2 data";
+      break;
+    case BZ_DATA_ERROR_MAGIC:
+      code = StatusCode::IOError;
+      ss << "data is not bz2-compressed (no magic header)";
+      break;
+    default:
+      code = StatusCode::UnknownError;
+      ss << "unknown bz2 error " << bz_result;
+      break;
+  }
+  return Status(code, ss.str());
+}
+
+// ----------------------------------------------------------------------
+// bz2 decompressor implementation
+
+class BZ2Decompressor : public Decompressor {
+ public:
+  BZ2Decompressor() : initialized_(false) {}
+
+  ~BZ2Decompressor() override {
+    if (initialized_) {
+      ARROW_UNUSED(BZ2_bzDecompressEnd(&stream_));
+    }
+  }
+
+  Status Init() {
+    DCHECK(!initialized_);
+    memset(&stream_, 0, sizeof(stream_));
+    int ret;
+    ret = BZ2_bzDecompressInit(&stream_, 0, 0);
+    if (ret != BZ_OK) {
+      return BZ2Error("bz2 decompressor init failed: ", ret);
+    }
+    initialized_ = true;
+    finished_ = false;
+    return Status::OK();
+  }
+
+  Status Decompress(int64_t input_len, const uint8_t* input, int64_t output_len,
+                    uint8_t* output, int64_t* bytes_read, int64_t* bytes_written,
+                    bool* need_more_output) override {
+    stream_.next_in = const_cast<char*>(reinterpret_cast<const char*>(input));
+    stream_.avail_in = static_cast<unsigned int>(std::min(input_len, kSizeLimit));
+    stream_.next_out = reinterpret_cast<char*>(output);
+    stream_.avail_out = static_cast<unsigned int>(std::min(output_len, kSizeLimit));
+    int ret;
+
+    ret = BZ2_bzDecompress(&stream_);
+    if (ret == BZ_OK || BZ_STREAM_END) {
```

Comment:
Is `ret ==` missing for `BZ_STREAM_END`?

Reply:
Good catch indeed!